secure_storage: introduce the secure storage subsystem

tomi-font opened the pull request:
Implements RFC https://github.com/zephyrproject-rtos/zephyr/issues/75275.

Generated documentation:
- [subsystem](https://builds.zephyrproject.io/zephyr/pr/76222/docs/services/secure_storage.html)
- [`persistent_key`](https://builds.zephyrproject.io/zephyr/pr/76222/docs/samples/psa/persistent_key/README.html) and [`psa_its`](https://builds.zephyrproject.io/zephyr/pr/76222/docs/samples/psa/its/README.html) samples
- [PSA Secure Storage API reference](https://builds.zephyrproject.io/zephyr/pr/76222/docs/doxygen/html/group__psa__secure__storage.html)
- [`secure_storage` internal API reference](https://builds.zephyrproject.io/zephyr/pr/76222/docs/doxygen/html/dir_b251feb5349caf21c27bf417dfd4e083.html)

Follow-up work could include:
- Implementing the ITS store module with a storage solution other than the settings subsystem.
- Establishing a way for different users of the PSA Crypto/Secure Storage APIs within Zephyr (subsystems, application) to use different ranges of numerical IDs.
- Giving the PSA PS API an implementation of its own for external flash support.
- Moving `tfm_integration` samples under `psa`, making them work both with and without TF-M.
- Making use of TF-M's PSA Secure Storage API tests to test Zephyr's implementation.

Below is a diagram of the subsystem's architecture in this PR.
Legend:
- Rectangle blocks indicate source code, while hexagon ones indicate header files.
- A full arrow means that a block always goes into the one it is pointing to, while a dotted one means that it may or may not depending on the configuration, meaning that different implementations (including custom ones) may be used.
```mermaid
%%{init: {"flowchart": {"htmlLabels": false}} }%%
graph TD;

psa_its{{"`**PSA ITS API**
<br>
psa_its_*()
_psa/
internal_trusted_storage.h_`"}}

psa_ps{{"`**PSA PS API**
<br>
psa_ps_*()
_psa/protected_storage.h_`"}}

its{{"`**ITS implementation**
<br>
secure_storage_its_*()
_its.h_`"}}

ps{{"`**PS implementation**
<br>
secure_storage_ps_*()
_ps.h_`"}}

psa_its-->its
psa_ps-.->its
psa_ps-.->ps

its_impl_zephyr["`Zephyr's ITS implementation
<br>
_its/implementation.c_`"]
ps_impl_custom["custom PS implementation"]

its-.->its_impl_zephyr
ps-->ps_impl_custom

its_transform{{"`**ITS transform module**
<br>
secure_storage_its_
transform_*()
_its/transform.h_`"}}

its_store{{"`**ITS store module**
<br>
secure_storage_its_
store_*()
_its/store.h_`"}}

its_impl_zephyr-->its_transform
its_impl_zephyr-->its_store

its_transform_aead["`AEAD ITS transform implementation
<br>
_its/transform/aead.c_`"]

its_store_settings["`settings ITS store implementation
<br>
_its/store_settings.c_`"]

its_transform-.->its_transform_aead

its_store-.->its_store_settings

its_transform_aead_api{{"`**AEAD ITS transform providers**
<br>
secure_storage_its_
transform_aead_get_*()
_its/transform/aead_get.h_`"}}

its_transform_aead-->its_transform_aead_api

its_transform_aead_api_default["`default AEAD scheme, key and nonce providers
<br>
_its/transform/aead_get.c_`"]

its_transform_aead_api-.->its_transform_aead_api_default
```

butok commented on samples/psa/persistent_key/README.rst:
```diff
@@ -0,0 +1,68 @@
+         :tool: west
+         :goals: build
+         :board: <ns_platform>
+         :west-args: -T sample.psa.persistent_key.tfm
```
Is using of -T mandatory?
What happens if to build it without the test item argument?

tomi-font replied:
It's not mandatory per se, it's just a convenient way to pick up the configuration. If you don't specify it you'll need to manually feed the needed Kconfig fragments as listed in the `sample.yaml` file depending on your build.

butok replied:
Samples should be buildable & functional without -T or other manual feeding.
Can it be a default sample configuration?
It is looks a bit strange that samples are using -T (test item). 
Probably you need to separate samples and tests.

tomi-font replied:
> Samples should be buildable & functional without -T or other manual feeding.

Is this a rule stated somewhere?

If the samples in this PR were to be buildable without configuration argument, there would need to be duplicated `.conf` files for every supported platform.
This would require explicitly listing the supported platforms, and duplicating the different possible configurations, which to me sounds really not good.
The `sample.yaml` file allows to nicely and automatically filter and feed the proper configuration.

The samples are default-buildable for NS platforms, but not for others because they require some more Kconfig options.

tomi-font replied:
> It looks a bit strange that samples are using -T (test item).
Probably you need to separate samples and tests.

`-T` in `west build` doesn't mean that it's a test, it just means (from `west build --help`): `Build based on test data in testcase.yaml or sample.yaml.`

butok replied:
> > Samples should be buildable & functional without -T or other manual feeding.
> 
> Is this a rule stated somewhere?
> 
> If the samples in this PR were to be buildable without configuration argument, there would need to be duplicated `.conf` files for every supported platform. This would require explicitly listing the supported platforms, and duplicating the different possible configurations, which to me sounds really not good. The `sample.yaml` file allows to nicely and automatically filter and feed the proper configuration.
> 
> The samples are default-buildable for NS platforms, but not for others because they require some more Kconfig options.

- --test-item is used as a build configuration. 
- the samples are tests.

It's better to wait for others' feedback.

butok replied:
> `-T` in `west build` doesn't mean that it's a test,

OK '-T' is named as "test-item" and points to a test data ;)

tomi-font replied:
> - --test-item is used as a build configuration.
> - the samples are tests.

You're right.
I am planning on adding actual tests, but they will be more unit tests to really push and test the functionality.
Those samples, even though they have a test aspect to them, demonstrate usage. So they do serve the purpose of a sample.
And I think that other samples elsewhere in Zephyr are also used as tests.

tomi-font replied:
Stumbled upon this: https://docs.zephyrproject.org/latest/samples/sample_definition_and_criteria.html#sample-criteria
To me the samples here do fulfill all the mentioned criteria.

butok replied:
> Stumbled upon this: https://docs.zephyrproject.org/latest/samples/sample_definition_and_criteria.html#sample-criteria To me the samples here do fulfill all the mentioned criteria.

Gray zone: "Samples must not be used as a testcase."

Do you still insist that the Zephyr sample **cannot** be functional without --test-item (as a required parameter)?
Do you know about other existing samples with this approach?
Ok to use --test-item for different optional test configurations, but a sample/test should work also with a default prj.conf.

tomi-font replied:
> Gray zone: "Samples must not be used as a testcase."

Below this statement is explicitly mentioned:

>Samples must not use Zephyr’s testing framework.
>
>    - Must not use [CONFIG_ZTEST](https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_ZTEST)
>
>    - Must not use zasserts in samples.
>
>If a sample can provide output that can be verified, then output should be evaluated against expected value to have some level of testing for the sample itself.

and a bit lower (in `2. Twister should be able to build every sample`):
> Sample output can be validated leveraging the harness_config regex option, wherever applicable.

<br>

> Do you still insist that the Zephyr sample cannot be functional without --test-item (as a required parameter)?

Not necessarily this parameter, as the alternative is to feed the Kconfig fragments to the build command directly (which is basically what `-T, --test-item` allows to do more conveniently).
But yes, I think it makes sense because it's way more scalable and maintainable when a sample has a few different configurations for all the platforms.

>  Do you know about other existing samples with this approach?

No.

> Ok to use --test-item for different optional test configurations, but a sample/test should work also with a default prj.conf.

Currently samples in this PR compile with the default configuration (no additional Kconfig fragment) on NS (TF-M-enabled) platforms.
It's possible to make them compile on all platforms by moving relevant configuration from Kconfig fragments to the `prj.conf` file, but that will output Kconfig warnings on NS platforms (because of trying to enable incompatible Kconfig options), so I don't think we want that.
The "proper" way would be to define platform-specific configurations in a `boards` directory. For that, you need to copy-paste the appropriate configuration for every platform you want to support.

So how would you go about doing this? What platforms would you define? And what would you then say for the platforms it would be possible to build for, but for which a configuration hasn't been manually defined?

And again:
> > Samples should be buildable & functional without -T or other manual feeding.
>
> Is this a rule stated somewhere?

And why does it seem so important if the sample's documentation clearly indicates how the sample should be built?

kartben replied:
> This would require explicitly listing the supported platforms, and duplicating the different possible configurations, which to me sounds really not good.

why? Why can't you list a couple of examples of the different "extra overlays" that need to be pulled in based on the flavor of board people are targetting?

tomi-font replied:
What do you mean exactly with `examples of the different "extra overlays"`? Do you mean documenting the possible build commands with `EXTRA_CONF_FILE`?

valeriosetti commented on subsys/secure_storage/src/its/transform/aead.c:
```diff
@@ -0,0 +1,122 @@
+	psa_set_key_algorithm(&key_attributes, alg);
+	psa_set_key_bits(&key_attributes, sizeof(key) * 8);
+
+	aead_crypt = (operation == PSA_KEY_USAGE_ENCRYPT) ?
```
Why are you calling directly into drivers' wrappers? IMO you should call `psa_aead_encrypt/psa_aead_decrypt` instead.

valeriosetti commented on subsys/secure_storage/src/psa/protected_storage.c:
```diff
@@ -0,0 +1,87 @@
+			psa_storage_create_flags_t create_flags)
+
+{
+#ifdef CONFIG_SECURE_STORAGE_PS_IMPLEMENTATION_CUSTOM
```
Instead of these `#ifdef` in C source files, why don't we use the same approach as for `aead.c/aead_get.c`, i.e. we include `protected_storage.c` only if `CONFIG_SECURE_STORAGE_PS_IMPLEMENTATION_CUSTOM` is not defined? IMO it would be nice to keep the same approach over the subsys.

valeriosetti commented on samples/psa/its/overlay-secure_storage.conf:
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_MBEDTLS=y
+CONFIG_MBEDTLS_PSA_CRYPTO_C=y
```
what's the purpose of Mbed TLS here? Unless I've missed something it's not used in the sample

valeriosetti reviewed (commented):
I just did a quick pass and I will review it again ASAP, but meanwhile I left some comments. 
Moreover I also has few other questions that I cannot add to the edited lines:

- can we name `subsys/secure_storage/src/its` as `subsys/secure_storage/src/secure_storage` instead? Because it's not only about ITS, as it contains implementation for the whole secure storage subsystem.

- I'd rename `aead_get.c` as `aead_get_key_nonce.c` because it looks clearer IMO

ceolin commented on include/zephyr/secure_storage/its.h:
```diff
@@ -0,0 +1,32 @@
+/* Copyright (c) 2024
```
This is confusing for me. Is this API supposed to be used directly or users should be using `psa_its_*` instead ?

AFAIU it is supposed to be the `psa`, so in this case why this header is public ? If it is not, shouldn't then it be agnostic and not use PSA types ? `psa_status_t`

ceolin reviewed (commented):
First of thanks for putting it together. It is a lot of work.

I have initially some questions about the architecture more than the implementation. Please look the comments.

valeriosetti commented on subsys/secure_storage/src/its/implementation.c:
```diff
@@ -0,0 +1,164 @@
+	ret = secure_storage_its_store_get(uid, sizeof(stored_data), stored_data, &stored_data_len);
+	if (ret != PSA_SUCCESS) {
+		if (ret != PSA_ERROR_DOES_NOT_EXIST) {
+			LOG_ERR("Failed to %s data %s storage. (%d)", "retrieve", "from", ret);
```
Nice idea. Then I suggest adding a `#define` for this at the top of this file in order to ensure the very same pattern is always used and prevent unintentional changes to some of these strings to break what you just explained.

str4t0m commented on subsys/secure_storage/Kconfig.its_transform:
```diff
@@ -0,0 +1,105 @@
+choice SECURE_STORAGE_ITS_TRANSFORM_AEAD_KEY_PROVIDER
+	prompt "AEAD ITS transform module encryption key provider"
+
+config SECURE_STORAGE_ITS_TRANSFORM_AEAD_KEY_PROVIDER_DEFAULT
```
Assuming, we might add additional implementations in the future. What do you think about renaming it more descriptively to something like SECURE_STORAGE_ITS_TRANSFORM_AEAD_KEY_PROVIDER_HWINFO

str4t0m commented on subsys/secure_storage/Kconfig.its_transform:
```diff
@@ -0,0 +1,105 @@
+	  processed as the key. In addition to the UID, it hashes the following, if available:
+	  - the device EUI64 as returned by hwinfo_get_device_eui64(), as first choice;
+	  - the device ID as returned by hwinfo_get_device_uuid(), as second choice.
+	  If no device ID is available or if the device ID is easily readable, the security of
+	  the key provided is low because the UIDs are easily guessable.
```
From what I understand, the uid is stored in plaintext(as additional data) so the hash could be calculated directly and `easily guessable` is not really true for the case without device id. We might modify that a bit, but haven't given it much thought yet.
Also, could we drop the implementation without hardware id, in case I'd open a PR to add an implementation for native posix, or would that be needed elsewhere?

tomi-font replied:
> From what I understand, the uid is stored in plaintext(as additional data)

No, it's not. It's used along with `create_flags` as additional data to authenticate, but the additional data is not part of the resulting ciphertext. You can see that `create_flags` is manually added to the `stored_entry` struct next to the ciphertext because we need to store it, but the UID is not found anywhere in the `stored_entry`.

(The reason why `create_flags` is not simply encrypted with the entry's data is that it would require using one more buffer, in which the data and `create_flags` would be put together, that would then be passed to PSA Crypto for the AEAD operation. It's avoided to be less hard on memory as the stack is already under heavy usage in there, and `create_flags` does not need to be kept secret. However we do want to authenticate both `create_flags` and the entry UID, thus they make up the additional data.)

Now, the entry UIDs actually do (may) get stored in plaintext by the store module. It needs to use them to be able to retrieve a particular entry. The current settings store module implementation directly uses the entry UIDs as setting names, so indeed `the key(hash) could be calculated directly`.

> and `easily guessable` is not really true for the case without device id. We might modify that a bit, but haven't given it much thought yet.

You're right. This description was left from an earlier version where I had encryption of the setting names (you can see it [here](https://github.com/tomi-font/zephyr/blob/e973b7714237d45d604107327cc7569a07ba0a70/subsys/secure_storage/src/its/storing_settings.c#L15-L64)). I used AES-ECB to encrypt the (single-block) entry UIDs used in the setting names with the same key used for encrypting the data when this insecure key provider was in use. That way, the UIDs were not actually stored in clear anywhere.
But because this was still not secure (just some more obfuscated) as the UIDs are easily guessable, I ended up removing it thinking it would be frowned upon. It could easily be brought back though.

> Also, could we drop the implementation without hardware id, in case I'd open a PR to add an implementation for native posix, or would that be needed elsewhere?

What would the implementation be like? And it would be only for simulation platforms?
The goal is to standardize on the PSA Secure Storage API, so it should be usable on all platforms, even if that means having only functional support (and no security) on certain ones. Even though the HW info API has drivers for a bunch of platforms, I don't think it's all and as such it's better to have a fallback solution for compatibility.

str4t0m replied:
> No, it's not. ... create_flags is manually added to the stored_entry.
Indeed, somehow I overlooked that and thought that both were added to the stored entry. I shouldn't have rushed the review submission in the morning. Sorry.

> , the entry UIDs actually do (may) get stored in plaintext by the store module. 
...
I used AES-ECB to encrypt the (single-block) entry UIDs used in the setting names with the same key used for encrypting the data when this insecure key provider was in use.

A few thoughts:
Looking at the Deployment models of arm's Security Risk Assessment of the API I think the current implementation goes into the direction of `DM.PROTECTED` but with no isolation. In case the the key is not derivable from the content of the storage medium, we could also support a deployment model similar to the mentioned `DM.EXPOSED`, but again without any isolation.
Additionally, in case a driver kernel would be implemented and user space enabled we would additionally be able to provide some isolation for user threads.

So believe we would need/(would be great) to define define the security objectives, requirements and threats for a few deployment scenarios we have in Zephyr to decide what protection mechanisms are needed and bring a benefit for each.

But that would probably take while. So I am curious about your timeline. I guess your priority is be to merge this soon with a basic implementation. To be able to convert the in-tree code and improve and provide additional implementations later on.

tomi-font replied:
> Looking at the Deployment models of arm's Security Risk Assessment of the API I think the current implementation goes into the direction of DM.PROTECTED but with no isolation. In case the the key is not derivable from the content of the storage medium, we could also support a deployment model similar to the mentioned DM.EXPOSED, but again without any isolation.

It's a bit hard as the end result will vary depending on what the end user chooses to do, but I'd say that right now the base implementation in this PR is more similar to `DM.EXPOSED`, without isolation. It's purely security at rest for the data.
The case where the key is derivable from the contents of the storage medium is really just for compatibility, and end users are urged to at least implement a secure key provider.

> Additionally, in case a driver kernel would be implemented and user space enabled we would additionally be able to provide some isolation for user threads.

Kernel and user space separation has already been mentioned, and it could be something to look into if we want to try to provide some sort of isolation.

> So believe we would need/(would be great) to define define the security objectives, requirements and threats for a few deployment scenarios we have in Zephyr to decide what protection mechanisms are needed and bring a benefit for each.

It could make sense. Especially if/when different implementations are made available (e.g. one where the storage medium is protected by the HW).

> But that would probably take while. So I am curious about your timeline. I guess your priority is be to merge this soon with a basic implementation. To be able to convert the in-tree code and improve and provide additional implementations later on?

Yeah, I think it would be good to get an initial, relatively basic implementation of the PSA Secure Storage API merged first, and to build on top of it. Having the API available in Zephyr will help other areas, one being the adoption of the PSA Crypto API.

str4t0m reviewed (dismissed):
Thanks for this efforts!
I have a few comments.

valeriosetti commented on modules/mbedtls/CMakeLists.txt:
```diff
@@ -161,6 +161,9 @@ zephyr_interface_library_named(mbedTLS)
     zephyr_library_sources(${crypto_source})
 
+    zephyr_library_compile_definitions(MBEDTLSCRYPTO_SOURCE_FILE)
+    zephyr_library_include_directories(internal_include)
```
> Even though right now it's only used for the ITS interface fed to Mbed TLS, I purposefully made the naming generic to describe what it actually means in practice.

But as of now (https://github.com/zephyrproject-rtos/zephyr/pull/76222/commits/52da601ccb1a546dbb36e734b4a0ddd6ede10c2c) `MBEDTLSCRYPTO_SOURCE_FILE` is used specifically to set `ITS_CALLER_ID` and not to select which head files to add...
Question: why cannot you just set `zephyr_library_compile_definitions(SECURE_STORAGE_ITS_CALLER_MBEDTLS)` in `zephyr/modules/mbedtls/CMakeLists.txt` and then check in `zephyr/subsys/secure_storage/include/psa/internal_trusted_storage.h` if that symbol is defined or not before assigning `ITS_CALLER_ID`? This avoid defining intermediate symbols and it would also solve the naming point above. Wdyt?

tomi-font replied:
Naming changed to `BUILDING_MBEDTLS_CRYPTO` to avoid potential confusion.

valeriosetti commented on subsys/secure_storage/src/its/transform/aead_get.c:
```diff
@@ -0,0 +1,114 @@
+	BUILD_ASSERT(CONFIG_SECURE_STORAGE_ITS_TRANSFORM_AEAD_KEY_SIZE <= HASH_OUTPUT_LEN);
+
+	if (CONFIG_SECURE_STORAGE_ITS_TRANSFORM_AEAD_KEY_SIZE == HASH_OUTPUT_LEN) {
+		return psa_hash_compute(PSA_ALG_SHA_256, (uint8_t *)&hash_input, hash_input_len,
```
Ok, thanks for the clarification. Perhaps my point-of-view is not too "embedded", but a SHA-256 is only 32 bytes so I didn't thought that buffer's length could be that bad.

tomi-font replied:
It wouldn't be too bad by itself, but the stack is already under heavy usage because of the other, bigger buffers, so any possible saving is good to take.

tomi-font commented on doc/services/secure_storage/index.rst:
```diff
@@ -0,0 +1,38 @@
+.. _secure_storage:
+
+secure storage
```
Will do. Please don't mind the documentation (other than that of the samples) too much for now, it's still a WIP.

tomi-font commented on samples/psa/persistent_key/src/main.c:
```diff
@@ -0,0 +1,130 @@
+		return -1;
+	}
+
+	ret = psa_destroy_key(SAMPLE_KEY_ID);
+	if (ret != PSA_ERROR_INVALID_HANDLE) {
+		LOG_ERR("Unexpected psa_destroy_key() return value when"
+			" destroying the key a second time. (%d)", ret);
+		return -1;
+	}
```
I will remove some testing/verification from the samples cause I indeed went a bit too far.

ceolin commented on include/zephyr/secure_storage/its.h:
```diff
@@ -0,0 +1,32 @@
+/* Copyright (c) 2024
```
Got the point. I think having these headers publicly available will cause people using them directly instead of psa* ones that are not in `include/zephyr`.  Maybe we should move them to `subsys/secure_storage/include` and only include them in the include path if the application sets `CONFIG_SECURE_STORAGE_ITS_IMPLEMENTATION_CUSTOM=y`. 

The way to use it is through PSA APIs so we need to make it clear, having a custom implementation is not the common use case so we don't need to promote these headers unless someone explicitly needs them.

What you think ?

tomi-font replied:
It makes sense.
I am working on adding CMake logic that will make `secure_storage` header files available only if the relevant `*_CUSTOM` Kconfig options are enabled.
It's a bit more complex than just conditionally adding an include path because there are different levels of granularity, but it's feasible.

tomi-font replied:
Solved as part of https://github.com/zephyrproject-rtos/zephyr/compare/37d6ae39bd509ee1f7a40f387f6a87a0fdfa3e4e..43bba9bf84ca53f3d134118b5d99f623602ce474#diff-ac1f1e035e1eb038ec993472202e54c6b07cd1d23aa0346426936e282c5ead8d.

Laczen commented:
@tomi-font, can this be a subsystem (no limitations on id etc)? Shouldn't this be part of modules/mbedtls?

tomi-font commented:
> can this be a subsystem

Why couldn't it?

> (no limitations on id etc)

I don't see what you're talking about? There are no such limitations.

> Shouldn't this be part of modules/mbedtls?

No, for at least the following reasons:
- PSA Secure Storage is not the same thing as PSA Crypto.
- And it's completely separate from Mbed TLS. It's not part of nor depends on Mbed TLS specifically in any way. It just happens to be using PSA Crypto (which could be provided by something else than Mbed TLS) for the AEAD operations in the base implementation, but that's an implementation detail.
- In fact, PSA Secure Storage is separate from Mbed TLS to the point where it could be used without Mbed TLS, because even if using the PSA Crypto API it could be implemented by something else than Mbed TLS (e.g. [wolfSSL](https://github.com/zephyrproject-rtos/zephyr/pull/75828)).

Laczen commented:
> > can this be a subsystem
> 
> Why couldn't it?
> 
> > (no limitations on id etc)
> 
> I don't see what you're talking about? There are no such limitations.
> 

Please correct me if I'm wrong: The proposed subsystem is also used directly by PSA-Crypto/MBEDTLS to store keys. If it is zephyr that defines what PSA-Crypto should use as id to store keys it is OK to use it as a subsystem, if OTOH it is PSA-Crypto that defines what is used as id to store it should be part of module/mbedtls (that includes PSA-Crypto)

It is also not a good idea from maintaining point of view to implement an externally defined API as a subsystem.

> > Shouldn't this be part of modules/mbedtls?
> 
> No, for at least the following reasons:
> 
> * PSA Secure Storage is not the same thing as PSA Crypto.
> * And it's completely separate from Mbed TLS. It's not part of nor depends on Mbed TLS specifically in any way. It just happens to be using PSA Crypto (which could be provided by something else than Mbed TLS) for the AEAD operations in the base implementation, but that's an implementation detail.
> * In fact, PSA Secure Storage is separate from Mbed TLS to the point where it could be used without Mbed TLS, because even if using the PSA Crypto API it could be implemented by something else than Mbed TLS (e.g. [wolfSSL](https://github.com/zephyrproject-rtos/zephyr/pull/75828)).

 I know that there is (will be) a clean separation between mbedtls and PSA-Crypto, but at the moment this is not yet true and module/mbedtls covers both mbedtls and PSA-Crypto.

The problem I'm seeing is not that the subsystem is using PSA-Crypto (and yes this can be changed to wolfSSL, ...), the problem is that the subsystem is **used by** PSA-Crypto for key storage and this might be out of control of zephyr.

Implementing PSA Secure Storage as part of module/mbedtls would still allow secure storage to be used by zephyr applications but it would need to follow the "rules" set by the configuration of PSA-Crypto (e.g. these are the id's that are allowed to be used) and any API changes would be handled by changes to module/mbedtls without influencing a zephyr subsystem.

@carlescufi, what is your opinion on adding an externally defined API as a subsystem ?

tomi-font commented on include/zephyr/secure_storage/its/transform.h:
```diff
@@ -0,0 +1,49 @@
+enum { SECURE_STORAGE_ITS_TRANSFORM_OUTPUT_SIZE
+	= sizeof(psa_storage_create_flags_t)
+	  + CONFIG_SECURE_STORAGE_ITS_MAX_DATA_SIZE
+	  + CONFIG_SECURE_STORAGE_ITS_TRANSFORM_OUTPUT_OVERHEAD };
```
One more advantage I just found out about is that it makes the generated API documentation of this header cleaner.
If `SECURE_STORAGE_ITS_TRANSFORM_OUTPUT_SIZE` is a macro, it gets expanded in the prototype of the functions that make use of it, which gives something rather ugly. When it's in an `enum` it shows up as is, it's not expanded.

carlescufi commented:
> @carlescufi, what is your opinion on adding an externally defined API as a subsystem ?

Are you talking about PSA when saying "externally defined API"? If so, there was a decision in the security working group to use PSA as the standard crypto API for Zephyr.

ceolin commented:
Hi @Laczen 

> > > can this be a subsystem
> > 
> > 
> > Why couldn't it?
> > > (no limitations on id etc)
> > 
> > 
> > I don't see what you're talking about? There are no such limitations.
> 
> Please correct me if I'm wrong: The proposed subsystem is also used directly by PSA-Crypto/MBEDTLS to store keys. If it is zephyr that defines what PSA-Crypto should use as id to store keys it is OK to use it as a subsystem, if OTOH it is PSA-Crypto that defines what is used as id to store it should be part of module/mbedtls (that includes PSA-Crypto)
> 
> It is also not a good idea from maintaining point of view to implement an externally defined API as a subsystem.
> 
> > > Shouldn't this be part of modules/mbedtls?
> > 
> > 
> > No, for at least the following reasons:
> > 
> > * PSA Secure Storage is not the same thing as PSA Crypto.
> > * And it's completely separate from Mbed TLS. It's not part of nor depends on Mbed TLS specifically in any way. It just happens to be using PSA Crypto (which could be provided by something else than Mbed TLS) for the AEAD operations in the base implementation, but that's an implementation detail.
> > * In fact, PSA Secure Storage is separate from Mbed TLS to the point where it could be used without Mbed TLS, because even if using the PSA Crypto API it could be implemented by something else than Mbed TLS (e.g. [wolfSSL](https://github.com/zephyrproject-rtos/zephyr/pull/75828)).
> 
> I know that there is (will be) a clean separation between mbedtls and PSA-Crypto, but at the moment this is not yet true and module/mbedtls covers both mbedtls and PSA-Crypto.
> 
> The problem I'm seeing is not that the subsystem is using PSA-Crypto (and yes this can be changed to wolfSSL, ...), the problem is that the subsystem is **used by** PSA-Crypto for key storage and this might be out of control of zephyr.

The usage is out of control ? Zephyr is implementing a well defined interface for that, as long as we do it properly (respecting all requirements) we should be good.
> 
> Implementing PSA Secure Storage as part of module/mbedtls would still allow secure storage to be used by zephyr applications but it would need to follow the "rules" set by the configuration of PSA-Crypto (e.g. these are the id's that are allowed to be used) and any API changes would be handled by changes to module/mbedtls without influencing a zephyr subsystem.

Another PSA Crypto implementation may require PSA Secure Storage. If the PSA Crypto "somehow" brings its own PSA secure storage we can simply disable it. The interface for the application is still the same.

> 
> @carlescufi, what is your opinion on adding an externally defined API as a subsystem ?

We are already doing it for crypto. Using something that is established increase portability. For example, if ones is already using these features with TF-M, changing to a target without TF-M won't require changes.

Laczen commented:
Hi @ceolin,

> 
> The usage is out of control ? Zephyr is implementing a well defined interface for that, as long as we do it properly (respecting all requirements) we should be good.
>

The problem is the requirements are unclear and they might change in future. However we can avoid this, see further on.
 
> > @carlescufi, what is your opinion on adding an externally defined API as a subsystem ?
> 
> We are already doing it for crypto. Using something that is established increase portability. For example, if ones is already using these features with TF-M, changing to a target without TF-M won't require changes.

I should have been more carefull in the wording: it should be read as: "What is your opinion on implementing an externally defined API as a subsystem". Are we already doing this for crypto, we are certainly using an externally defined api inside subsystems, but is an implementation done as a subsystem?

When we add an external library to zephyr we are forking this external library and adding some glue logic to the fork to allow use of zephyr drivers/subsystems when required. PSA-Crypto needs some glue logic in order to use zephyr for persistent storage of keys (and mbedtls is also using this for storage of other items). This glue logic would implement `psa_its_get(), ...`. In TF-M this this glue logic is not needed because PSA-Crypto can use the TF-M provided implementations.

Now the next step is we would like to "use" the API for securely storing application items. It is not possible to use all id's because some of them might already be defined by PSA-Crypto glue logic or the TF-M implementation. We can "import" the externally defined ranges, but they might change. An alternative would be to let the module define what range we can use, IMHO this would be the simplest: let TF-M or the glue logic define what id's can be used by a zephyr subsystem for secure storage (e.g. reserve a start id and range for application storage). The zephyr subsystem could then "use" (not implement) the TF-M or PSA-Crypto glue logic to provide the secure storage.

ceolin commented:
> Hi @ceolin,
> 
> > The usage is out of control ? Zephyr is implementing a well defined interface for that, as long as we do it properly (respecting all requirements) we should be good.
> 
> The problem is the requirements are unclear and they might change in future. However we can avoid this, see further on.
> 
> > > @carlescufi, what is your opinion on adding an externally defined API as a subsystem ?
> > 
> > 
> > We are already doing it for crypto. Using something that is established increase portability. For example, if ones is already using these features with TF-M, changing to a target without TF-M won't require changes.
> 
> I should have been more carefull in the wording: it should be read as: "What is your opinion on implementing an externally defined API as a subsystem". Are we already doing this for crypto, we are certainly using an externally defined api inside subsystems, but is an implementation done as a subsystem?
> 
> When we add an external library to zephyr we are forking this external library and adding some glue logic to the fork to allow use of zephyr drivers/subsystems when required. PSA-Crypto needs some glue logic in order to use zephyr for persistent storage of keys (and mbedtls is also using this for storage of other items). This glue logic would implement `psa_its_get(), ...`. In TF-M this this glue logic is not needed because PSA-Crypto can use the TF-M provided implementations.
> 

That is a good point indeed but from my perspective there are some points that makes it different. The secure storage is not a fork from another project, it is a Zephyr implementation of the specification (in this case the API). Make it part of PSA Crypto / MBEDTLS fork is not a good thing too since we should keep them as close as possible from the original.

tomi-font commented:
Fixed the reference in the release notes entry and rebased to fix CI failures.

valeriosetti commented on tests/subsys/secure_storage/psa/its/prj.conf:
```diff
@@ -0,0 +1,17 @@
+CONFIG_TEST_RANDOM_GENERATOR=y
+CONFIG_TIMER_RANDOM_GENERATOR=y
+CONFIG_MBEDTLS_ENTROPY_POLL_ZEPHYR=y
+CONFIG_MBEDTLS_PSA_CRYPTO_C=y
```
I understand that the purpose of this code is to showcase that the internal storage proposed here is 100% compatible with the one provided by TF-M, but building for a TF-M plaform (ex: `nrf5340dk/nrf5340/cpuapp/ns` as listed in the `testcase.yaml`) gives a warning that
```
warning: MBEDTLS_PSA_CRYPTO_C (defined at modules/mbedtls/Kconfig.tls-generic:514,
modules/mbedtls/Kconfig.tls-generic:514) was assigned the value 'y' but got the value 'n'
```

Wouldn't it be nicer to split the case in `testcase.yaml` in order to have a configuration for TF-M and one for non-TF-M devices? So that both builds without warnings?

However I don't have a strong opinion on this, so if you don't agree just set this one as solved with a comment ;)

tomi-font replied:
Yeah, good point. I was aware of this, but as this single Kconfig fragment works for all board targets, I was like let's do it like this, I guess because of the previous discussion regarding this topic on the samples. And I thought that we wouldn't care about compilation warnings in tests, but it's true that it's not very clean.
I did what you suggested.

valeriosetti reviewed (commented):
Overall looks good to me. I just left some minor comment

tomi-font commented on subsys/secure_storage/src/its/transform/aead.c:
```diff
@@ -0,0 +1,128 @@
+} __packed;
+
+psa_status_t secure_storage_its_transform_to_store(
+			secure_storage_its_uid_t uid, size_t data_len, const void *data,
+			psa_storage_create_flags_t create_flags, uint8_t stored_data
+			[static SECURE_STORAGE_ITS_TRANSFORM_MAX_STORED_DATA_SIZE],
+			size_t *stored_data_len)
```
Actually the pattern is uniform, in line with the PSA Secure Storage API specification. It's as follows, in order:
- the current `size` (if the buffer is receiving data) or `length` (if the buffer already contains data) of the buffer
- the buffer
- if the buffer is receiving data, the output `length` i.e. how many bytes were placed in the buffer (so the argument is a pointer)

In this case there is no `size` because it's embedded as part of the buffer argument, between the square brackets. And `stored_data_len` is the output length.

valeriosetti replied:
> in line with the PSA Secure Storage API specification

You are right. For example https://arm-software.github.io/psa-api/storage/1.0/api/api.html#c.psa_its_set sets `length` before `p_data`. Sorry for the comment then. I got confused by the fact that the convention for the crypto is different. In that case (ex: https://arm-software.github.io/psa-api/crypto/1.2/api/ops/sign.html#c.psa_sign_message) there's `input` and then `input_length`.

Anyway, this is not an issue then. Solved.

tomi-font replied:
Yeah, I saw that. A bit too bad that the two APIs are not aligned on this.

de-nordic commented on subsys/secure_storage/src/its/transform/aead_get.c:
```diff
@@ -0,0 +1,141 @@
+{
+	size_t hash_len;
+
+#if CONFIG_SECURE_STORAGE_ITS_TRANSFORM_AEAD_KEY_SIZE == SHA256_OUTPUT_SIZE
```
Just a note: this could done by `if` as compiler is able to optimize dead code when sees comparison of literals.

de-nordic commented on subsys/secure_storage/src/its/implementation.c:
```diff
@@ -0,0 +1,228 @@
+		return true;
+	}
+	if (existing_data_len == data_length
+	 && existing_create_flags == create_flags
+	 && !memcmp(existing_data, p_data, data_length)) {
```
Discussed offline, just not my personal choice of style.

tomi-font commented:
Wrote some more documentation on ways the implementation purposefully deviates from the official specification.

de-nordic reviewed (dismissed):
The code using settings looks OK.
With the acknowledgement of limitations in security of the storage provided, as stated in design and documentation, the support for the PSA API looks ok.

Laczen commented:
> @tomi-font I am still having problems with how this is added, but as said before, I might be wrong.
> 
> A. If the subsystem is used without any psa-crypto use for key storage, the zephyr application can use any id to store data, B. If the subsystem is used with psa-crypto for key storage, the zephyr application can only use id's below `PSA_KEY_ID_USER_MIN`, but this can't be enforced.
> 
> It would be good to add some tests to make sure that there is no misuse of the used id in all supported configuration options.

No answer ? As said I might be wrong, but please provide an answer.

tomi-font commented:
> No answer ? As said I might be wrong, but please provide an answer.

No answer because indeed you are wrong and the answer was already in the code before your original comment. Just as the nice-to-have [ID test](https://github.com/zephyrproject-rtos/zephyr/commit/558424863f2469bef081d56cff82e0b4ad5d9ba0#diff-a26d8b0a95a0971539ca888d302c423a7027872cb4a56bb4c64866e14fcc8907R32) you were talking about.

kartben commented:
> Just as the nice-to-have [ID test](https://github.com/zephyrproject-rtos/zephyr/commit/558424863f2469bef081d56cff82e0b4ad5d9ba0#diff-a26d8b0a95a0971539ca888d302c423a7027872cb4a56bb4c64866e14fcc8907R32) you were talking about.

That commit message needs some love though :) not sure about the others but might be worth doing a quick check

tomi-font commented:
> That commit message needs some love though :) not sure about the others but might be worth doing a quick check

Not sure what you mean here. Do you mean that the commit should have a longer description?

kartben commented:
> > That commit message needs some love though :) not sure about the others but might be worth doing a quick check
> 
> Not sure what you mean here. Do you mean that the commit should have a longer description?

The commit message for the commit you linked above literally said "Implements <to be filled>". The latest update is marginally better but still has placeholder text, which doesn't look good for such a large change. See also my previous comment that I just edited

tomi-font commented:
> The commit message for the commit you linked above literally said `"Implements <to be filled>"`. The latest update is marginally better but still has placeholder text, which doesn't look good for such a large change. See also my previous comment that I just edited

In my above comment I linked to a particular commit because I was pointing to a particular file and line. I want the link to remain valid even after I push more changes to this PR, thus I took the changes from an earlier push (also to show that the changes were already present back then).

The placeholder is because I am planning to link to the RFC in the commit message, but when I do this every time I push changes to the PR there is [a notification](https://github.com/zephyrproject-rtos/zephyr/issues/75275#ref-commit-d37959a) posted on the RFC. I wanted to avoid the spam, thus the placeholder until the PR is closer to being mergeable.

> would be good to put more actual meat in the commit message itself rather than point to external resources

Do you consider the RFC, that is an issue in this repository, to be an external resource?
I am also pointing to the documentation I wrote, which is part of the commit. Not external in my opinion. If I were to write more stuff in the commit message, I would just duplicate what is in the documentation, and in a less exhaustive manner. Which is why I would rather avoid doing that, and point people to the actual documentation directly.

Laczen commented:
> > No answer ? As said I might be wrong, but please provide an answer.
> 
> No answer because indeed you are wrong and the answer was already in the code before your original comment. Just as the nice-to-have [ID test](https://github.com/zephyrproject-rtos/zephyr/commit/558424863f2469bef081d56cff82e0b4ad5d9ba0#diff-a26d8b0a95a0971539ca888d302c423a7027872cb4a56bb4c64866e14fcc8907R32) you were talking about.

OK, you could just have pointed me to that test.

Anyhow, I still think it is not good to place this in the zephyr tree. The correct place would be under the module that provides the definitions of `psa_xxx` types.

tomi-font commented:
Added some more tests for custom module implementations and an optimization on the stored `create_flags` to save 3 bytes per entry.

de-nordic commented:
@str4t0m I have dismissed you old review, please re-review.

de-nordic commented on tests/subsys/secure_storage/psa/its/testcase.yaml:
```diff
@@ -0,0 +1,32 @@
+  integration_platforms:
+    - native_sim
+  platform_exclude:
+    - qemu_cortex_m0 # settings subsystem initialization fails
```
The qemu_cortex_m0 seems broken, or at least not read data from flash.bin.
The fash0 dts definition comes from nordic  nrf51822, which causes all flash_reads on internal flash (flash0) to be done as memcpy of some memory region to target buffer, ending up with some trash in user provided buffers (this makes NVS fail), instead of using the flash.bin.

str4t0m reviewed (commented):
Thanks, with the constraints clearly documented my concerns have been addressed.
Just a question regarding device-testing.

tomi-font commented on tests/subsys/secure_storage/psa/its/testcase.yaml:
> I guess that a couple of platforms that have a too large write bock size for nvs will fail device-testing with the current filters or am I overlooking something? Would be care?

Why would they fail? With the filtering implemented [here](https://github.com/zephyrproject-rtos/zephyr/pull/76222/files/2e6bc955f98468bffd862550b4ebb04d010d070b#diff-fb5682f6af21bb48291e3d8c7e6c7a2ba99a9601dfc2e97f4deabe459424350dR10-R12) (at the Kconfig level) I tried to filter out board targets that wouldn't properly support settings (currently the only storage backend). Though it's possible that not all of them are caught (see `qemu_cortex_m0` that was just discussed), especially at run time. I did try locally a `twister --all --build-only`, though.
Maybe @de-nordic has something to add about your question?

> And regarding test_write_once_flag this should be fine when CI is run with --west-flash="--erase"

Yes, it is fine in that scenario. I just wanted to clearly document this in case someone runs this test locally (or somehow doesn't erase the flash between runs).

str4t0m replied:
It think it would be just boards that define a storage partition and return an error during init.
But those wouldn't block CI anyways.

> Yes, it is fine in that scenario. I just wanted to clearly document this in case someone runs this test locally (or somehow doesn't erase the flash between runs).

Just wanted to assure myself, should have phrased that a bit different.

nashif reviewed (changes requested):
can this PR be split into multiple commits please? this iway too much to go into 1 commit: docs, subsys, integration, tests, samples.....

tomi-font committed:
secure_storage: introduce the secure storage subsystem

Implements RFC https://github.com/zephyrproject-rtos/zephyr/issues/75275.

See also the PR (https://github.com/zephyrproject-rtos/zephyr/pull/76222)
for more information.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
modules: mbedtls: enable persistent keys when CONFIG_SECURE_STORAGE

With the secure storage subsystem now providing an implementation of
the PSA ITS API, let Mbed TLS use it when it's enabled.
This allows the use of persistent keys in the PSA Crypto API.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
tests: secure_storage: add tests for the secure storage subsystem

Add one suite to test the PSA ITS API directly, and another to test it
through the PSA Crypto API.

The PSA ITS API test suite is run with different configurations to also
test customization possibilities that the secure storage subsystem offers.

Both test suites also have a test scenario with TF-M instead of the
secure storage subsystem.
This serves as compatibility testing to ensure that the PSA Secure
Storage API behaves the same from the user's point of view regardless
of the underlying implementation provider.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
samples: psa: its: add the psa_its sample

Add a sample to demonstrate direct use of the PSA ITS API.

The implementation of the API is provided either by the just-introduced
secure storage subsystem or by TF-M.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
samples: psa: persistent_key: add the persistent_key sample

Add a sample to demonstrate use of persistent keys in the PSA Crypto API.

The implementation of the PSA ITS API that allows storage of persistent
keys is provided either by the just-introduced secure storage subsystem
or by TF-M.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
doc: secure_storage: add documentation for the secure storage subsystem

It includes both the high-level documentation of the subsystem and
that of APIs defined in code (PSA Secure Storage and internal APIs of
the subsystem).

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font commented:
> can this PR be split into multiple commits please? this iway too much to go into 1 commit: docs, subsys, integration, tests, samples.....

Done! By the way, I added an `area: Secure storage` label in the `MAINTAINERS.yml` entry, taking example on the existing ones. I imagine this will require to manually create a new GitHub label?

tomi-font commented:
ping @ceolin @nashif

tomi-font commented:
> Yep, I just created it @tomi-font

Thanks!
